### PR TITLE
TLDS: Add 1D sampler

### PR DESCRIPTION
**Pull request by raven02**
Used by Mario Rabbids 

>[  17.769891] HW.GPU <Critical> video_core\renderer_opengl\gl_shader_decompiler.cpp:CompileInstr:2027: Unhandled texture type 0
>[  17.802382] Render.OpenGL <Error> video_core\renderer_opengl\gl_shader_util.cpp:LoadShader:44: Error compiling vertex shader:
0(384) : error C1008: undefined variable "coords"

### CODE REVIEW

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 1]
```diff
@@ -1987,13 +1987,18 @@ class GLSLGenerator {
...
-                ASSERT(instr.tlds.IsArrayTexture() == false);
                 std::string coord;
+                Tegra::Shader::TextureType texture_type{instr.tlds.GetTextureType()};
+                bool is_array{instr.tlds.IsArrayTexture()};
 
-                switch (instr.tlds.GetTextureType()) {
+                switch (texture_type) {
+                case Tegra::Shader::TextureType::Texture1D: {
+                    const std::string x = regs.GetRegisterAsInteger(instr.gpr8);
+                    coord = "int coords = " + x + ';';
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Did you verify that this is actually the register used for this sampling?

File: src/video_core/renderer_opengl/gl_rasterizer_cache.h [hunk 2]
```diff
@@ -121,6 +121,7 @@ struct SurfaceParams {
     static SurfaceTarget SurfaceTargetFromTextureType(Tegra::Texture::TextureType texture_type) {
         switch (texture_type) {
         case Tegra::Texture::TextureType::Texture1D:
+        case Tegra::Texture::TextureType::Texture1DBuffer:
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.h]: We still have no idea how Texture1DBuffer works, remove this code

Comment (raven02) [reply src/video_core/renderer_opengl/gl_rasterizer_cache.h]: Alright.

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp (hunk 1 above)
Comment (raven02) [reply src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: No , probably need more work to verify it.

File: src/video_core/renderer_opengl/gl_rasterizer_cache.h (hunk 2 above)
Comment (Subv) [reply src/video_core/renderer_opengl/gl_rasterizer_cache.h]: there exists GL_TEXTURE_BUFFER but i still think we need to do more research before we even attempt this.

Comment (raven02) [reply src/video_core/renderer_opengl/gl_rasterizer_cache.h]: Alright Subv

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 3]
```diff
@@ -1987,13 +1987,18 @@ class GLSLGenerator {
                 break;
             }
             case OpCode::Id::TLDS: {
-                ASSERT(instr.tlds.GetTextureType() == Tegra::Shader::TextureType::Texture2D);
-                ASSERT(instr.tlds.IsArrayTexture() == false);
                 std::string coord;
+                Tegra::Shader::TextureType texture_type{instr.tlds.GetTextureType()};
```
[the remark this answers is not in the document]
Comment (raven02) [reply src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Will do for TLDS and others instr.

**Comment by raven02** [comment]
Addressed comments from Subv and Lioncash

**Comment by FernandoS27** [comment]
The coordinate does seem to be correct.

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 4]
```diff
@@ -1987,13 +1987,18 @@ class GLSLGenerator {
...
-                switch (instr.tlds.GetTextureType()) {
+                switch (texture_type) {
+                case Tegra::Shader::TextureType::Texture1D: {
+                    const std::string x = regs.GetRegisterAsInteger(instr.gpr8);
+                    coord = "int coords = " + x + ';';
+                    break;
+                }
                 case Tegra::Shader::TextureType::Texture2D: {
-                    if (instr.tlds.IsArrayTexture()) {
+                    if (is_array) {
```
Comment (bunnei) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Shouldn't this be checked outside of the case? e.g. for non-2D array textures?

**Comment by bunnei** [comment]
@raven02 can you please address my above comment and rebase? Thanks!

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp (hunk 4 above)
Comment (raven02) [reply src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Sorry for late . I basically refactor it however i agree this check can be done outside of the case .

**Comment by raven02** [comment]
Accordingly to https://envytools.readthedocs.io/en/latest/hw/memory/g80-surface.html

1D Buffer is actually a packed 1D texture array of elements

File: src/video_core/renderer_opengl/gl_rasterizer_cache.h [hunk 5]
```diff
@@ -122,6 +122,7 @@ struct SurfaceParams {
...
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.h]: I still do not like this, I much prefer we defer the implementation until we have a proper test for it (Maybe a piglit test covers this?)

Comment (raven02) [reply src/video_core/renderer_opengl/gl_rasterizer_cache.h]: Done

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 6]
```diff
@@ -2018,9 +2018,9 @@ class GLSLGenerator {
                 break;
             }
             case OpCode::Id::TLDS: {
-                ASSERT(instr.tlds.GetTextureType() == Tegra::Shader::TextureType::Texture2D);
-                ASSERT(instr.tlds.IsArrayTexture() == false);
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: You removed the array assert, add it back, then LGTM

Comment (raven02) [reply src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Done

**Pull request merged**